### PR TITLE
SNOW-878116 Add support for PARTITION BY to COPY INTO location

### PR DESCRIPTION
**Pull request by sfc-gh-lcalderonachio**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes https://github.com/snowflakedb/snowflake-sqlalchemy/issues/430

Co-authored by @azban 

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

This adds an argument to CopyInto which can be used to specify the PARTITION BY value.

### CODE REVIEW

**Comment by sfc-gh-lcalderonachio** [comment]
I have read the CLA Document and I hereby sign the CLA

File: src/snowflake/sqlalchemy/base.py [hunk 1]
```diff
@@ -576,6 +576,19 @@ def visit_copy_into(self, copy_into, **kw):
...
+        partition_by_value = ""
+        if isinstance(copy_into.partition_by, (BindParameter, Executable)):
+            partition_by_value = copy_into.partition_by.compile(
+                compile_kwargs={"literal_binds": True}
+            )
+        elif copy_into.partition_by is not None:
+            partition_by_value = copy_into.partition_by
+
+        partition_by = (
+            f"PARTITION BY {partition_by_value}" if partition_by_value != "" else ""
```
[the remark this answers is not in the document]
Comment (sfc-gh-lcalderonachio) [reply src/snowflake/sqlalchemy/base.py]: Thanks for the observations here. I have updated the code to make partition_by_value None by default and check if the partition_by should be an empty string or not based on this.
Right now, if the partition_by_value variable is None, then partition_by will be an empty string, effectively not adding anything related at all to the PARTITION BY to the COPY INTO statement.
The existing test case test_copy_into_location already validates this in most of the asserts where they do not have the PARTITION BY clause, only in the new added ones.

[the remark this answers is not in the document]
Comment (sfc-gh-lcalderonachio) [reply src/snowflake/sqlalchemy/base.py]: I have added an extra test case where the empty string case is being validated https://github.com/snowflakedb/snowflake-sqlalchemy/pull/542/commits/5179d3efbd951839e02d1b1f8961c8847bc663a6

File: DESCRIPTION.md [hunk 2]
```diff
@@ -19,6 +19,7 @@ Source code is also available at:
   - Add support for iceberg table with Snowflake Catalog
   - Fix cluster by option to support explicit expressions
   - Add support for MAP datatype
+  - Add support for partition by to copy into <location>
```
Comment (sfc-gh-jvasquezrojas) [inline DESCRIPTION.md:22]: Add this to Unreleased section

Comment (sfc-gh-lcalderonachio) [reply DESCRIPTION.md]: Thanks. Changed in https://github.com/snowflakedb/snowflake-sqlalchemy/pull/542/commits/300a4fb672984309698c059840e6aaf24193268a

**Review by sfc-gh-jvasquezrojas — APPROVED** [review]
